XamarinCommunityToolkit/ObjectModel/AsyncCommand.shared.cs

```diff
@@ -0,0 +1,155 @@
+		public AsyncCommand(
+			Func<T, Task> execute,
+			Func<T, bool> canExecute = null,
+			Action<Exception> onException = null,
+			bool continueOnCapturedContext = false)
+		{
+			this.execute = execute ?? throw new ArgumentNullException(nameof(execute));
+			this.canExecute = canExecute;
+			this.onException = onException;
+			this.continueOnCapturedContext = continueOnCapturedContext;
+		}
+
+		/// <summary>
+		/// Invoke the CanExecute method and return if it can be executed.
+		/// </summary>
+		/// <param name="parameter">Parameter to pass to CanExecute.</param>
+		/// <returns>If it can be executed.</returns>
+		public bool CanExecute(object parameter)
+		{
+			if (!IsValidParameter(parameter))
+				return false;
+			if (IsExecuting && !AllowMultipleExecution)
+				return false;
+
+			return canExecute?.Invoke((T) parameter) ?? true;
+		}
+
+		/// <summary>
+		/// Event triggered when Can Execute changes.
+		/// </summary>
+		public event EventHandler CanExecuteChanged
+		{
+			add => weakEventManager.AddEventHandler(value);
+			remove => weakEventManager.RemoveEventHandler(value);
+		}
+
+		public async void Execute(object parameter) => await ExecuteAsync((T) parameter);
+
+		/// <summary>
+		/// Execute the command async.
+		/// </summary>
+		/// <returns>Task of action being executed that can be awaited.</returns>
+		public async Task ExecuteAsync(T parameter)
+		{
+			try
+			{
+				IsExecuting = Interlocked.Increment(ref executionCount) > 0;
+				await execute(parameter).ConfigureAwait(continueOnCapturedContext);
+			}
+			catch (Exception e) when (onException != null)
+			{
+				onException(e);
+			}
+			finally
+			{
+				IsExecuting = Interlocked.Decrement(ref executionCount) > 0;
+			}
+		}
+
+		/// <summary>
+		/// Raise a CanExecute change event.
+		/// </summary>
+		public void RaiseCanExecuteChanged() =>
+			weakEventManager.HandleEvent(this, EventArgs.Empty, nameof(CanExecuteChanged));
+
+		static bool IsValidParameter(object o)
+		{
+			if (o != null)
+			{
+				// The parameter isn't null, so we don't have to worry whether null is a valid option
+				return o is T;
+			}
+
+			var t = typeof(T);
+
+			// The parameter is null. Is T Nullable?
+			if (Nullable.GetUnderlyingType(t) != null)
+			{
+				return true;
+			}
+
+			// Not a Nullable, if it's a value type then null is not valid
+			return !t.GetTypeInfo().IsValueType;
+		}
+	}
+
+	public class AsyncCommand : AsyncCommand<object>
+	{
+		public AsyncCommand(Func<Task> execute, Func<bool> canExecute = null, Action<Exception> onException = null,
+			bool continueOnCapturedContext = false) : base(o => execute(), o => canExecute?.Invoke() ?? true,
+			onException, continueOnCapturedContext)
+		{
```

Comment:
This is a useless check cuz the same check is in the base class

Reply:
I don't think so, because we pass a lambda expression to the base class. I added a unit test to check this.

Reply:
You are right)
Maybe consider making this method's body as lambda as well?)

```
=> if (execute == null)
     throw new ArgumentNullException(nameof(execute));
```

Reply:
we can put a `if` statement in lambda?

Reply:
good point ))